src/tbb/queuing_rw_mutex.cpp

```diff
@@ -458,7 +494,7 @@ struct queuing_rw_mutex_impl {
         } // if( this != my_mutex->q_tail... )
         {
             unsigned char old_state = STATE_UPGRADE_REQUESTED;
-            s.my_state.compare_exchange_strong(old_state, STATE_UPGRADE_WAITING, std::memory_order_acquire);
+            s.my_state.compare_exchange_strong(old_state, STATE_UPGRADE_WAITING, std::memory_order_release, std::memory_order_acquire);
```

Comment:
```suggestion
            // If we reach STATE_UPGRADE_WAITING state we do not build happens-before relation with READER on
            // left. We delegate this responsibility to READER on left when it try upgrading. Therefore, we are releasing
            // on success.
            // Otherwise, on fail, we already acquired the next->my_state.
            s.my_state.compare_exchange_strong(old_state, STATE_UPGRADE_WAITING, std::memory_order_release, std::memory_order_relaxed);
```